src/gui/csvImport/CsvImportWidget.cpp

```diff
@@ -0,0 +1,289 @@
+
+    for (int c=0; c<m_ui->tableViewFields->horizontalHeader()->count(); ++c) {
+        m_ui->tableViewFields->horizontalHeader()->setSectionResizeMode(
+                    c, QHeaderView::Stretch);
+    }
+}
+
+void CsvImportWidget::updatePreview() {
+
+    m_ui->labelSizeRowsCols->setText(m_parserModel->getFileInfo());
+    m_ui->spinBoxSkip->setValue(0);
+    m_ui->spinBoxSkip->setMaximum(m_parserModel->rowCount()-1);
+
+    int i;
+    QStringList list(tr("Not present in CSV file"));
+
+    for (i=1; i<m_parserModel->getCsvCols(); i++) {
+        QString s = QString(tr("Column ")) + QString::number(i);
+        list << s;
+    }
+    m_comboModel->setStringList(list);
+
+    i=1;
+    Q_FOREACH (QComboBox* b, m_combos) {
+        if (i < m_parserModel->getCsvCols()) {
+            b->setCurrentIndex(i);
+        }
+        else {
+            b->setCurrentIndex(0);
+        }
+        ++i;
+    }
+}
+
+void CsvImportWidget::load(const QString& filename, Database* const db) {
+    m_db = db;
+    m_parserModel->setFilename(filename);
+    m_ui->labelFilename->setText(filename);
+    Group* group = m_db->rootGroup();
+    group->setUuid(Uuid::random());
+    group->setNotes(tr("Imported from CSV file\nOriginal data: ") + filename);
+
+    parse();
+}
+
+void CsvImportWidget::parse() {
+    configParser();
+    QApplication::setOverrideCursor(Qt::WaitCursor);
+    bool good = m_parserModel->parse();
+    QApplication::restoreOverrideCursor();
+    updatePreview();
+    m_ui->pushButtonWarnings->setEnabled(!good);
+}
+
+void CsvImportWidget::showReport() {
+    MessageBox::warning(this, tr("Syntax error"), tr("While parsing file...\n")
+               .append(m_parserModel->getStatus()), QMessageBox::Ok, QMessageBox::Ok);
+}
+
+void CsvImportWidget::writeDatabase() {
+
+    checkGroupNames();
+    for (int r=0; r<m_parserModel->rowCount(); r++) {
+        //use the validity of second column as a GO/NOGO hint for all others fields
+        if (m_parserModel->data(m_parserModel->index(r, 1)).isValid()) {
+            Entry* entry = new Entry();
+            entry->setUuid(Uuid::random());
+            entry->setGroup(splitGroups(m_parserModel->data(m_parserModel->index(r, 0)).toString()));
+            entry->setTitle(    m_parserModel->data(m_parserModel->index(r, 1)).toString());
+            entry->setUsername( m_parserModel->data(m_parserModel->index(r, 2)).toString());
+            entry->setPassword( m_parserModel->data(m_parserModel->index(r, 3)).toString());
+            entry->setUrl(      m_parserModel->data(m_parserModel->index(r, 4)).toString());
+            entry->setNotes(    m_parserModel->data(m_parserModel->index(r, 5)).toString());
+        }
+    }
+
+    QBuffer buffer;
+    buffer.open(QBuffer::ReadWrite);
+
+    KeePass2Writer writer;
+    writer.writeDatabase(&buffer, m_db);
+    if (writer.hasError()) {
+        MessageBox::warning(this, tr("Error"), tr("CSV import: writer has errors:\n")
```

Comment:
@phoerious Would not use messageWidget here because subsequent `emit editFinished` prevents the dialog to show up